### PR TITLE
Add s3tokens validation.

### PR DESCRIPTION
**Pull request by chmouel**
There is still some works to be done in the swift middleware but this
should works.

Documentation and tests are in the planning.

### CODE REVIEW

File: keystone/contrib/s3/core.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+from hashlib import sha1
+from keystone import catalog
+from keystone import config
+from keystone import identity
+from keystone import policy
+from keystone import token
+from keystone import service
+from keystone.common import wsgi
+from keystone.contrib.ec2 import Manager as EC2Manager
```
Comment (termie) [inline keystone/contrib/s3/core.py]: should import the module not the class, so:

from keystone.contrib import ec2

and later use ec2.Manager()

File: keystone/contrib/s3/core.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        # infos.
+
+
+class S3Controller(wsgi.Application):
+    def __init__(self):
+        self.catalog_api = catalog.Manager()
+        self.identity_api = identity.Manager()
+        self.token_api = token.Manager()
+        self.policy_api = policy.Manager()
+        self.ec2_api = EC2Manager()
```
Comment (termie) [inline keystone/contrib/s3/core.py]: ec2.Manager()

File: keystone/contrib/s3/core.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        :param context: standard context
+        :param credentials: dict of s3 signature
+        :returns: token: openstack token equivalent to access key along
+                         with the corresponding service catalog and roles
+        """
+
+        creds_ref = self.ec2_api.get_credential(context,
+                                                credentials['access'])
+
+        msg = base64.urlsafe_b64decode(str(credentials['token']))
```
Comment (termie) [inline keystone/contrib/s3/core.py:32]: could you make this little section a separate "sign_s3" function?

File: keystone/contrib/s3/core.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                                              creds_ref['user_id'])
+        metadata_ref = self.identity_api.get_metadata(
+                context=context,
+                user_id=user_ref['id'],
+                tenant_id=tenant_ref['id'])
+        catalog_ref = self.catalog_api.get_catalog(
+                context=context,
+                user_id=user_ref['id'],
+                tenant_id=tenant_ref['id'],
+                    metadata=metadata_ref)
```
Comment (termie) [inline keystone/contrib/s3/core.py]: extra indent

File: keystone/contrib/s3/core.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+class S3Controller(wsgi.Application):
+    def __init__(self):
+        self.catalog_api = catalog.Manager()
+        self.identity_api = identity.Manager()
+        self.token_api = token.Manager()
+        self.policy_api = policy.Manager()
+        self.ec2_api = EC2Manager()
+        super(S3Controller, self).__init__()
+
+    def authenticate_s3(self, context, credentials=None):
```
Comment (termie) [inline keystone/contrib/s3/core.py]: could this whole function wrap authenticate_ec2 and just change the signing? As in, make the signing part an instance method and have the s3 class override the signer

**Comment by anotherjesse** [comment]
I've never liked the "ec2 token" name because the tokens are both AWS-style tokens....

**Comment by chmouel** [comment]
What about having a /awstokens for tokens CRUD and /{ec2,s3}tokens for
specific services validation? everything will be under
keystone.contrib.awstokens  directory.

File: keystone/contrib/ec2/core.py [hunk 6]
```diff
@@ -83,7 +85,24 @@ def __init__(self):
...
+            credentials['host'] = hostname
+            signature = signer.generate(credentials)
+            if signature != credentials.signature:
+                # TODO(termie): proper exception
+                raise Exception("Not Authorized")
+        else:
+            raise Exception("Not Authorized")
+
+    def authenticate(self, context, credentials=None,
                          ec2Credentials=None):
```
Comment (termie) [inline keystone/contrib/ec2/core.py:106]: could you re-indent this to match with self, pls

File: keystone/contrib/ec2/core.py [hunk 7]
```diff
@@ -162,8 +171,9 @@ def authenticate_ec2(self, context, credentials=None,
         # TODO(termie): i don't think the ec2 middleware currently expects a
         #               full return, but it contains a note saying that it
         #               would be better to expect a full return
-        return TokenController._format_authenticate(
-                self, token_ref, roles_ref, catalog_ref)
+        token_controller = service.TokenController()
+        return token_controller._format_authenticate(
+            token_ref, roles_ref, catalog_ref)
```
Comment (termie) [inline keystone/contrib/ec2/core.py]: when moving to a new line for a long line, typically one double-indents the second line (as it was before this change)

File: keystone/contrib/ec2/core.py [hunk 8]
```diff
@@ -113,27 +132,17 @@ def authenticate_ec2(self, context, credentials=None,
...
-            raise Exception("Not Authorized")
+        self.check_signature(creds_ref, credentials)
 
         # TODO(termie): don't create new tokens every time
         # TODO(termie): this is copied from TokenController.authenticate
         token_id = uuid.uuid4().hex
-        tenant_ref = self.identity_api.get_tenant(creds_ref['tenant_id'])
-        user_ref = self.identity_api.get_user(creds_ref['user_id'])
+        tenant_ref = self.identity_api.get_tenant(
+            context=context,
```
Comment (termie) [inline keystone/contrib/ec2/core.py]: could you double-indent or align these with the end of get_tenant(

File: keystone/contrib/ec2/core.py (hunk 7 above)
Comment (chmouel) [inline keystone/contrib/ec2/core.py]: > when moving to a new line for a long line, typically one double-indents
> the second line (as it was before this change)

emacs python-mode doesn't seem to like it much , oh well I'll use vim for
that one..

Comment (termie) [inline keystone/contrib/ec2/core.py]: weird, pretty sure that one is in pep8's guide too

Comment (chmouel) [inline keystone/contrib/ec2/core.py]: > weird, pretty sure that one is in pep8's guide too
> 
> FWIW: bin/pep8 command (0.6.1) doesn't seem to bug about it, but well
> anyway I have updated the commit ranges of this pull req with the
> indentations as they were before.

**Pull request merged**